hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/BaseSparkInternalRowReaderContext.java

```diff
@@ -0,0 +1,126 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi;
+
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.engine.HoodieReaderContext;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.HoodieEmptyRecord;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordMerger;
+import org.apache.hudi.common.model.HoodieSparkRecord;
+import org.apache.hudi.common.util.ConfigUtils;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.exception.HoodieException;
+
+import org.apache.avro.Schema;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.spark.sql.HoodieInternalRowUtils;
+import org.apache.spark.sql.HoodieUnsafeRowUtils;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.Map;
+
+import static org.apache.hudi.common.model.HoodieRecord.RECORD_KEY_METADATA_FIELD;
+import static org.apache.hudi.common.model.HoodieRecordMerger.DEFAULT_MERGER_STRATEGY_UUID;
+
+/**
+ * An abstract class implementing {@link HoodieReaderContext} to handle {@link InternalRow}s.
+ * Subclasses need to implement {@link #getFileRecordIterator} with the reader logic.
+ */
+public abstract class BaseSparkInternalRowReaderContext extends HoodieReaderContext<InternalRow> {
+  @Override
+  public FileSystem getFs(String path, Configuration conf) {
+    return FSUtils.getFs(path, conf);
+  }
+
+  @Override
+  public HoodieRecordMerger getRecordMerger(String mergerStrategy) {
+    switch (mergerStrategy) {
+      case DEFAULT_MERGER_STRATEGY_UUID:
+        return new HoodieSparkRecordMerger();
+      default:
+        throw new HoodieException("The merger strategy UUID is not supported: " + mergerStrategy);
+    }
+  }
+
+  @Override
+  public Object getValue(InternalRow row, Schema schema, String fieldName) {
+    return getFieldValueFromInternalRow(row, schema, fieldName);
+  }
+
+  @Override
+  public String getRecordKey(InternalRow row, Schema schema) {
+    return getFieldValueFromInternalRow(row, schema, RECORD_KEY_METADATA_FIELD).toString();
+  }
+
+  @Override
+  public Comparable getOrderingValue(Option<InternalRow> rowOption,
+                                     Map<String, Object> metadataMap,
+                                     Schema schema,
+                                     TypedProperties props) {
+    if (metadataMap.containsKey(INTERNAL_META_ORDERING_FIELD)) {
+      return (Comparable) metadataMap.get(INTERNAL_META_ORDERING_FIELD);
+    }
+
+    if (!rowOption.isPresent()) {
+      return 0;
```

Comment:
Curious: 0 is the smallest value?

Reply:
This condition should not be hit, as either the `INTERNAL_META_ORDERING_FIELD` exists or the row exists. Here's is more like a placeholder following similar logic in the `HoodieSparkRecord#getOrderingValue` when ordering field is not defined so natural ordering is followed.

```
  @Override
  public Comparable<?> getOrderingValue(Schema recordSchema, Properties props) {
    StructType structType = HoodieInternalRowUtils.getCachedSchema(recordSchema);
    String orderingField = ConfigUtils.getOrderingField(props);
    scala.Option<NestedFieldPath> cachedNestedFieldPath =
        HoodieInternalRowUtils.getCachedPosList(structType, orderingField);
    if (cachedNestedFieldPath.isDefined()) {
      NestedFieldPath nestedFieldPath = cachedNestedFieldPath.get();
      return (Comparable<?>) HoodieUnsafeRowUtils.getNestedInternalRowValue(data, nestedFieldPath);
    } else {
      return 0;
    }
  }
```